### PR TITLE
DVC-7981 manual event tracking

### PR DESCRIPTION
**Pull request by chris-hoefgen**
- add support for the client.track() call
- track variable default event when the client is not initialized
- attempt to trigger a flush when manual event is logged but he queue is full
- more unit tests

### CODE REVIEW

File: devcycle_python_sdk/managers/event_queue_manager.py [hunk 1]
```diff
@@ -114,4 +125,68 @@ def run(self):
...
+
+        event_json = json.dumps(event.to_json())
+        if bucketed_config and bucketed_config.variable_variation_map:
+            variation_map_json = json.dumps(bucketed_config.variable_variation_map)
+        else:
+            variation_map_json = "{}"
+        self._local_bucketing.queue_aggregate_event(event_json, variation_map_json)
+
+    def _check_queue_status(self) -> None:
+        if self._flush_needed() and not self._flush_lock.locked():
```
Comment (rob-odwyer) [inline devcycle_python_sdk/managers/event_queue_manager.py:180]: I think this is a redundant check of `_flush_lock.locked()`? It'll just check it in `_flush_events` anyways.

File: devcycle_python_sdk/models/event.py [hunk 2]
```diff
@@ -18,22 +18,26 @@ class EventType:
...
     def to_json(self):
-        return {
+        json_dict = {
             key: getattr(self, key)
             for key in self.__dataclass_fields__
-            if getattr(self, key) is not None
+            if getattr(self, key) is not None and key != "date"
         }
+        if self.date:
+            json_dict["date"] = f"{self.date.isoformat()}Z"
```
Comment (rob-odwyer) [inline devcycle_python_sdk/models/event.py:32]: This date field is user-provided, right? Is there a risk here that they provide us a non-UTC datetime, and we ignore the timezone and use UTC?

File: devcycle_python_sdk/managers/event_queue_manager.py [hunk 3]
```diff
@@ -114,4 +125,68 @@ def run(self):
...
+        if event is None:
+            raise ValueError("event cannot be None")
+
+        if not event.type:
+            raise ValueError("event type not set")
+
+        try:
+            self._check_queue_status()
+        except QueueFullError:
+            logger.warning(f"Event queue is full, dropping event: {event}")
```
[the remark this answers is not in the document]
Comment (chris-hoefgen) [reply devcycle_python_sdk/managers/event_queue_manager.py:147]: good call

File: devcycle_python_sdk/dvc_cloud_client.py [hunk 4]
```diff
@@ -154,3 +138,23 @@ def track(self, user: User, user_event: Event) -> None:
...
+
+def _validate_sdk_key(sdk_key: str) -> None:
+    if sdk_key is None or len(sdk_key) == 0:
+        raise ValueError(
+            "Missing environment key! Call initialize with a valid environment key"
+        )
+
+    if not sdk_key.startswith("server") and not sdk_key.startswith("dvc_server"):
+        raise ValueError(
+            "Missing environment key! Call initialize with a valid environment key"
```
[the remark this answers is not in the document]
Comment (chris-hoefgen) [reply devcycle_python_sdk/dvc_cloud_client.py:151]: i flipped back and forth.  It seems the message is different in different tests and in some cases both works.  

```
expectErrorMessageToBe(
                body.exception,
                'Missing environment key! Call initialize with a valid environment key',
                'Invalid SDK key provided. Call build with a valid server SDK key',
            )
```

File: devcycle_python_sdk/models/event.py (hunk 2 above)
Comment (chris-hoefgen) [reply devcycle_python_sdk/models/event.py:32]: hmm, fair point.  i didn't get a straight answer out of adam as to what should be used for timezone.  Maybe the right thing to do is convert it to UTC and then format it?

File: devcycle_python_sdk/dvc_cloud_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rob-odwyer) [reply devcycle_python_sdk/dvc_cloud_client.py:151]: Yeah, I think I added the extra option in the test harness there to cut down on the warnings 😆 

I think the second message there is more "correct" though - if you had it set, but to something wrong, saying it was missing might be a bit confusing.

File: devcycle_python_sdk/managers/event_queue_manager.py (hunk 1 above)
Comment (chris-hoefgen) [reply devcycle_python_sdk/managers/event_queue_manager.py:180]: true.  i'll remove the extra check

File: devcycle_python_sdk/dvc_cloud_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chris-hoefgen) [reply devcycle_python_sdk/dvc_cloud_client.py:151]: i agree. We should try to just make it all SDK Key.  i'll change it

File: devcycle_python_sdk/models/event.py (hunk 2 above)
Comment (rob-odwyer) [reply devcycle_python_sdk/models/event.py:32]: Probably, yeah. We might need the `pytz` library for that, from what I remember `datetime` just provides the attribute for a timezone but not the actual values?

Unless we can define one for UTC ourselves?

Comment (rob-odwyer) [reply devcycle_python_sdk/models/event.py:32]: Just read up on it some more - I think as long as we only want to support converting to UTC and not other more exotic timezones, we can do something like this:
```
from datetime import datetime, timezone, timedelta
...
dt.astimezone(timezone.utc)
```
And then if we receive a `datetime` with a timezone defined by some other library (pytz, etc) we'll still be able to convert it.

Comment (chris-hoefgen) [reply devcycle_python_sdk/models/event.py]: ok, i added a conversion - just force it before changing it to json

**Review by rob-odwyer — APPROVED** [review]

**Pull request merged**